HACKING.adoc

```diff
@@ -0,0 +1,254 @@
+  tag
+  http://caml.inria.fr/mantis/search.php?project_id=1&sticky_issues=1&sortby=last_updated&dir=DESC&highlight_changed=24&hide_status_id=90&tag_string=junior_job[junior_job].
+
+* The
+  https://github.com/ocamllabs/compiler-hacking/wiki/Things-to-work-on[OCaml
+  Labs compiler-hacking wiki] contains various ideas of changes to
+  propose, some easy, some requiring sensibly more work.
+
+* Documentation improvements are always most appreciated, either in
+  the various mli files or in the official manual
+  (See link:manual/README.md[]). If you invest effort in understanding
+  a part of the codebase, submitting a pull request that adds
+  clarifying comment can be an excellent contribution to help future
+  you and your peers.
+
+* The https://github.com/ocaml/ocaml[github project] contains a lot of
+  pull requests, many of them being in dire need of a review -- we
+  have more people willing to contribute changes than to review
+  someone else's change. Picking one of them, trying to understand the
+  code (looking at the code around it) and asking questions about what
+  you don't understand or what feels odd is super-useful. It helps the
+  contribution process, and it is also an excellent way to get to know
+  various parts of the compiler from the angle of a specific aspect or
+  feature.
+
+   Again, reviewing small or medium-sized pull requests is accessible
+  to anyone with OCaml programming experience, and helps maintainers
+  and other contributors. If you also submit pull requets yourself,
+  a good discipline to follow is to review at least as many pull
+  requests as you submit.
+
+== Structure of the compiler
+
+The compiler code base can be intimidating at first sight. Here are
+a few pointers to get started.
+
+=== Compilation pipeline
+
+==== The driver -- link:driver/[]
+
+The driver contains the "main" function of the compilers that drive
+compilation. It is in charge of parsing the command-line arguments,
+and then composing the compiler passes according to the invoked
+command-line, by calling functions from the various parts of the
+compiler described below.
+
+==== Parsing -- link:parsing/[]
+
+Parses source files and produces an Abstract Syntax Tree (AST)
+(link:parsing/parsetree.mli[] has lot of helpful comments). See
+link:parsing/HACKING.adoc[].
+
+The logic for Camlp4 and Ppx preprocessing is not in link:parsing/[],
+but in link:driver/[], see link:driver/pparse.mli[],
+link:driver/pparse.mli[].
+
+==== Typing -- link:typing/[]
+
+Type-checks syntactic AST and produces a typed representation
+(link:parsing/typedtree.mli[] has some helpful comments). See
+link:typing/HACKING.adoc[].
+
+==== The bytecode compiler -- link:bytecomp/[]
+
+==== The native compiler -- link:middle_end/[] and link:asmcomp/[]
+
+=== Runtime system
+
+=== Libraries
+
+link:stdlib/[]:: The standard library. Each file is mostly independent and should not need further knowledge.
```

Comment:
> Is it worth mentioning the internal modules (especially the modules related to Format, etc.) straightaway?

This would be interesting but should go in `stdlib/HACKING.adoc` I think, in the interest of keeping the central file concise and focused on generic, subsystem-agnostic information advice.